k3s-helper: Log certificate on error.

mook-as opened the pull request:
This makes an extra log for certificate details when failing to fetch k3s releases &c; this is meant to troubleshoot user issues.

dforste commented:
This didn't properly debug what was happening. The fetch.log is present but empty. These are the only log files with any content. 
[update.log](https://github.com/rancher-sandbox/rancher-desktop/files/8017006/update.log)
[background.log](https://github.com/rancher-sandbox/rancher-desktop/files/8017009/background.log)

mook-as commented:
Made improvements:
- The error message now goes to the corresponding log file, rather than `fetch.log`.
- We will print out the details of the whole failing certificate chain (in case we spot the MITM CA in there).
- We also handle the updater.

At this point, I think we can (assuming code review passes) land this in the app to get more details in general.

ericpromislow commented on src/utils/fetch.ts:
```diff
@@ -0,0 +1,136 @@
+    }
+  }
+
+  [util.inspect.custom](depth: number, opts: util.InspectOptionsStylized) {
```
What is the `[util.inspect.custom]` syntax? Is this generating a method name on the fly, based on the value of `util.inspect.custom` at class loading time? That doesn't really make sense to me.

ericpromislow commented on src/utils/fetch.ts:
```diff
@@ -0,0 +1,136 @@
+          if (options.agent instanceof http.Agent) {
+            agent = options.agent;
+          } else {
+            agent = options.agent(parsedURL);
```
What would `options.agent` be in this case? Is this set as a function in our code base or somewhere else?

ericpromislow commented on src/utils/fetch.ts:
```diff
@@ -0,0 +1,136 @@
+
+        // Need to construct a custom agent.
+        const secureAgent = agent as https.Agent ?? https.globalAgent;
+        let ca = secureAgent.options.ca ?? [];
```
Not thrilled that `ca` stands for `custom agent` here when we're adjacent to code where it stands for `cert authority`.  And if that's what it stands for, maybe the var name should be plural, like `customAgents`.

ericpromislow commented on src/utils/fetch.ts:
```diff
@@ -0,0 +1,136 @@
+let systemCerts: string[];
+
+/**
+ * A wrapper for node-fetch to log certificates on error.
```
There's more going on here -- could you provide a longer comment for maintainers?

mook-as replied:
Added more comments.

While adding comments, I realized that we should keep as much of the existing Agent as possible (since that broke proxies); so I wrapped it instead. Sorry for the large code churn…

ericpromislow commented:
And is there a good way to repro an error condition? I tried munging the SystemCerts but didn't see anything in the logs.

mook-as commented:
To reproduce:
- Change [k3s channels URL](https://github.com/rancher-sandbox/rancher-desktop/blob/14adc4e2f060b24bd87edb4eaff40a2cdb0b28e8/src/k8s-engine/k3sHelper.ts#L58) to use `https://expired.badssl.com` instead. (There are other `badssl.com` subdomains to test other cert failures.)
- Remove `~\AppData\Local\rancher-desktop\cache\k3s-versions.json` (or other correct path on your system)
- `npm run dev`.

ericpromislow reviewed (commented):
Thanks for the repro. I assume we have nothing to do with the error message, given that the heading is "A JavaScript error occurred in the main process" and the message is `TypeError: Cannot read properties of undefined (reading 'fingerprint')
at new CertificateVerificationError` with an even less useful traceback.

Should we catch that in the `fetch` wrapper and emit a friendlier error message.

mook-as commented:
> Thanks for the repro. I assume we have nothing to do with the error message, given that the heading is "A JavaScript error occurred in the main process" and the message is `TypeError: Cannot read properties of undefined (reading 'fingerprint') at new CertificateVerificationError` with an even less useful traceback.
> 
No, that sounds like an error I fixed in f6e2a478f065396ea989635587417c6f546e7ab3… — specifically [here](https://github.com/rancher-sandbox/rancher-desktop/pull/1487/commits/f6e2a478f065396ea989635587417c6f546e7ab3#diff-6825020adff99694c861ec029be8e69f0e3c638320bcdb2445d526c1bc619c6cR24) where I added a `?` when reading off `issuerCertificate`.  If you're using the newest code and still seeing the issues, please see what else you can dig up?  If there's any tracebacks around `new CertificateVerificationError` that would be helpful.

ericpromislow reviewed (approved):
Looks better with the recent change.
